src/mcx_core.cu

```diff
@@ -1385,6 +1385,60 @@ __device__ inline int launchnewphoton(MCXpos* p, MCXdir* v, Stokes* s, MCXtime*
                     canfocus = (gcfg->srctype == MCX_SRC_SLIT);
                     break;
                 }
+                case (MCX_SRC_RING): {
+                    /* ring light source, can be a complete ring or a truncated ring. the expected parameters for definition of the
+                    light source are as follows:
+                    gcfg->srcparam1.x: minimum ring radius in mm
+                    gcfg->srcparam1.y: maximum ring radius in mm
+                    gcfg->srcparam1.z: first angular range in degrees. Defines the angular width of the first ring section
+                    gcfg->srcparam1.w: second angular range in degrees. Defines the angular width of the second ring section
+                    gcfg->srcparam2.x: full beam divergence in degrees. 0 means photons are launched straight down. angle greater than
+                    0 means that the photons are launched with a direction sampled from a normal distribution with FWHM of half the
+                    beam divergence.
+                    for simulating a complete ring the parameters should be defined as: gcfg->srcparam1.z = 180 and gcfg->srcparam1.w = 180
+                    */
+                    float minimum_radius = gcfg->srcparam1.x;
+                    float maximum_radius = gcfg->srcparam1.y;
+                    float cut_angle_1 = (TWO_PI / 360.f) * gcfg->srcparam1.z;
+                    float cut_angle_2 = (TWO_PI / 360.f) * gcfg->srcparam1.w;
```

Comment:
can you minimize the use of registers (temporary variables) in the new code? the current biggest bottleneck to mcx's GPU kernel is the number of registers - the total register used by mcx is ranging between 64 to 100+ (for more complex modes such as SVMC).

you can see other source definitions - we tried to minimize the use of temporary variables and compute temporary values on the fly or reuse some existing ones.

this new source alone added 24 new registers, which amounts to nearly half of the registers used by the barebone mcx kernel. I am pretty sure merging this code to the main code will significantly slow down the rest of the functions as the compiler is not smart enough to optimize out these registers.

Reply:
I understand that minimize the use of registers will hurt readability, but speed performance is the most important spec of mcx

Reply:
this makes me thinking if you can simply merge the angle range to the existing disk/ring source code? likely this will yield the same result but much more compact implementation. you can still call this ring source but just add a new `case` statement in this section

https://github.com/fangq/mcx/blob/1a2f29118ee0ff25597853fd765c802d76ce330a/src/mcx_core.cu#L1232